### DISCUSSION TITLE
New organ : problem on loading

### DISCUSSION THREAD

**Discussion by Jean-Gabriel1982** [GrandOrgue/grandorgue#680] [General]
Hello,

I composed an organ with some different organs available for GrandOrgue and Hauptwerk.
I used as usual the Vincent Forman's Organ builder program.

When my organ was finished, i tested it from Vincent Forman's prrogram.
Everything was ok.

But when i load it on GrandOrgue (version v0.3.1.2341-15.os), i have this message

> 16/09/2021 12:41:58: 12:41:58: Erreur : Valeur hors limite section 'Organ' entrée 'NumberOfRanks': 443

I don't understand what happened and don't know how to correct that ..
Have you an idea ?

Many thanks :)

**Comment by rousseldenis** [comment]
@larspalo Is there a limit range to Ranks amount for an Organ?

**Reply by larspalo** [reply]
Yes, according to the help it's 400. I'll look into the sources and see if I find the exact files.

**Reply by Jean-Gabriel1982** [reply]
@larspalo 

Ah ... oups .. is there a way to increase the limit ?
I made a so nice organ :p ...

**Reply by larspalo** [reply]
Are you able to build GO yourself? If so, then this is an excellent opportunity for you to test changing ./src/grandorgue/GOrgueModel.cpp:101:	m_ODFRankCount = cfg.ReadInteger(ODFSetting, group, wxT("NumberOfRanks"), 0, 400, false); the 400 value to for instance 999 and build from the source and test.

**Reply by Jean-Gabriel1982** [reply]
No unfortunately I don't know how to code at all and am unable to build ...  :(

**Reply by larspalo** [reply]
Why not learn? It's not that difficult you know!

**Reply by larspalo** [reply]
We'll help you and you help us!

**Reply by Jean-Gabriel1982** [reply]
Oh I will be happy to help! I'm a fan of GrandOrgue and super happy that this fantastic program is evolving!
But I know absolutely nothing about the code and am more afraid of doing stupid things ...

**Reply by larspalo** [reply]
No problem, everyone can learn. I don't know what OS you're using but development is easiest on Linux. If you're using Windows, do you have any spare computer or possibility to setup a VM (like VirtualBox or something similar)?

**Reply by Jean-Gabriel1982** [reply]
I'm on Linux Mint :)

**Reply by larspalo** [reply]
Ok, then it's pretty easy! You'll need some basic development packages first.

1. Install gcc C++ compiler, make, cmake and the development packages of wxWigets, jack (libjack), pkg-config, fftw (fftw3), wavpack and alsa (libasound) from your distribution - that is all the -dev packages mentioned.
2. Install docbook-xsl, xsltproc, zip, gettext and po4a (if present on your distribution)

I usually use Synaptic to install the packages, but you can use whatever method you're comfortable with.

Then the next step will be to get the GO source. For that you might want to install the git package. Then you can issue the command:
git clone --recurse-submodules https://github.com/GrandOrgue/grandorgue.git
that will download the source into a grandorgue folder.
The next steps can be done in different ways, but I describe what I do. In the grandorgue folder and create a new folder called build-for, then in that folder create a folder called LinuxRelease. Then in that folder you issue the command:
cmake -DCMAKE_BUILD_TYPE=Release -G "Unix Makefiles" ../..
If you don't get any errors you can follow up with:
make
and wait until it's finished.
The application (GrandOrgue) will be found in the bin folder of your LinuxRelease folder.

If you get this far, then it's time to learn how to modify the source files. For this issue you should open the src/grandorgue/GOrgueModel.cpp file and find line 101 that reads:
m_ODFRankCount = cfg.ReadInteger(ODFSetting, group, wxT("NumberOfRanks"), 0, 400, false);
and change the 400 value to 999. Save the file, go back to the LinuxRelease folder and rerun make to create your first modified build of GrandOrgue.

If you get any trouble just chime in and we'll find out how to fix it. Good luck!

I really think I'll make a video of this process to show just how easy it is to do...

**Reply by oleg68** [reply]
> 
> 
> Ah ... oups .. is there a way to increase the limit ?

@larspalo I think the best way of increasing the limit is *to create an issue here*.

GO users are obligated neigther to be programmers nor to know how to build GO from sources.

**Reply by rousseldenis** [reply]
Yes indeed, the good way for contributing is to point to issues. Then, on a second step is to do a PR (and people involved can be different).

Sometimes, an issue is not necessary if the change is short and obvious.

**Reply by larspalo** [reply]
Of course, but friends - you're missing an opportunity! It's something we could fix in a blink of an eye, but what's the fun in that?

**Reply by rousseldenis** [reply]
> Of course, but friends - you're missing an opportunity! It's something we could fix in a blink of an eye, but what's the fun in that?

Hehe. A path to an online GrandOrgue event ?

How to contribute? Inside of GrandOrgue What's cool in GO?

Cool topics!

**Reply by larspalo** [reply]
Let's face it. We need more people willing to contribute not only issues but also possible solutions. If someone is interested in learning - I'm all for helping them and pointing them in the right direction. That's how I was recruited by others who believed in me...

**Reply by rousseldenis** [reply]
Indeed, I'm listening too as I'm a beginner at this time (already did the build steps). Let's go forward.

**Reply by rousseldenis** [reply]
By the way, I'm aware of a solution for an alternative of OrganBuilder on linux (I haven't tested it with Wine but I like having all the tools locally).

**Reply by Jean-Gabriel1982** [reply]
> > Ah ... oups .. is there a way to increase the limit ?
> 
> @larspalo I think the best way of increasing the limit is _to create an issue here_.
> 
> GO users are obligated neigther to be programmers nor to know how to build GO from sources.

I will try …. I want to try … but I dont promise you any success :D

I think if it's easy to correct you will have done better to do it it will be much faster than me hehe

but ultimately, if it is something that is not a priority, I can try to do it at my own pace.

it's not that I don't like coding but I imagine myself a mountain to climb and I'm afraid of not being productive :-)

**Reply by Jean-Gabriel1982** [reply]
> By the way, I'm aware of a solution for an alternative of OrganBuilder on linux (I haven't tested it with Wine but I like having all the tools locally).

That would be great !!

**Reply by oleg68** [reply]
I've created a pull request #682 

> Let's face it. We need more people willing to contribute not only issues but also possible solutions. If someone is interested in learning - I'm all for helping them and pointing them in the right direction. That's how I was recruited by others who believed in me

Agree. But building GO from source is too complex for @Jean-Gabriel1982. I'd start learning with something easier, for example with code review.

@rousseldenis could you help to add @Jean-Gabriel1982 to #682  as a reviewer?

**Reply by rousseldenis** [reply]
I've added him to members team. As soon as he accepts the invitation, you can add him to reviewers

**Comment by larspalo** [comment]
[Oleg](https://github.com/oleg68) just made a PR with the changes I suggested, so it will be available soon to the world. However, @Jean-Gabriel1982 , I hope that you will test building GO yourself anyway and see how easy it is! You can do it! The step is smaller than you think...

**Comment by oleg68** [comment]
@Jean-Gabriel1982 You can test this changing from https://github.com/oleg68/GrandOrgue-oficial/actions/runs/1242891624

**Comment by oleg68** [comment]
Reading [BUILD.md](https://github.com/GrandOrgue/grandorgue/blob/master/BUILD.md) may also be useful.

But the simplest way is not to build youself but to allow GitHub to build GO after making changes.

**Comment by oleg68** [comment]
@Jean-Gabriel1982 Have you tried the new intermediate version?

**Reply by Jean-Gabriel1982** [reply]
@oleg68 I'm sorry I didn't have the time because of very long days at work :-(
I had planned to watch this evening ... but precisely I don’t know how to install this intermediate version ...

**Reply by oleg68** [reply]
If you have a release version been already installed (with dependencies), then the best option for an intermediate version is to try it without installation: download the x86_64.tar.gz file, unpack it to some directory and run GrandOrgue from the bin subdirectory. See INSTALL.md for details.

After a next release is issued with this change, you will be able to downlosd .deb files and update to the new release with the same apt-get command.

**Reply by Jean-Gabriel1982** [reply]
No I have some troubles now with my screens and the problem will be corrected this evening. Then I will install GO. Right now it is not installed at all. 

But at work, I have Windows 10 and installed the 0.3.xx-15 (from your fork) and maybe I can install the intermediate version and test ? I can do it this morning

**Reply by Jean-Gabriel1982** [reply]
> 
> 
> If you have a release version been already installed (with dependencies), then the best option for an intermediate version is to try it without installation: download the x86_64.tar.gz file, unpack it to some directory and run GrandOrgue from the bin subdirectory. See INSTALL.md for details.
> 
> After a next release is issued with this change, you will be able to downlosd .deb files and update to the new release with the same apt-get command.

I downloaded the version v0.3.1.2342-1 for Windows, but i think the Number of Ranks was not changed in this version, so it does not work.

**Reply by oleg68** [reply]
You can find intermediation versions at the 'Actions' tab on the official grandorgue page.

Of cause you can try a new version without installing on windows too. For this you have to download an artifact ending with .windows.x86_64.zip and do the same as on linux. Attention: all intermediate artifacts from github come zipped, so you need to unzip it twice.

**Reply by Jean-Gabriel1982** [reply]
Ok i try without installing, and i don't have any message regarding NumberOfRanks.

But i have this message, duplicated for each pipe (so many many many lines :) )

> Warning : name of file (pipe.wav) incompatible with upper / lower case sensitive systems

EDIT : and the Organ does not load ..
For closing GO i had to kill it via ctrl+alt+suppr :p

**Reply by oleg68** [reply]
It is a warning, not error. After the medsages the organ should load.

But for compatibility with linux it is berrer to correct either the sample file names or their names in the organ file.

**Reply by Jean-Gabriel1982** [reply]
> 
> 
> It is a warning, not error. After the medsages the organ should load.

I can't load the entire organ as i have just 8Gb Ram and i think my organ needs at least 40Gb Ram.. I'll check on linux this evening

> But for compatibility with linux it is berrer to correct either the sample file names or their names in the organ file.

Yes but i can't correct 8000 files like that :O 
When i make the organ with Mr Forman programm, everything is ok ... i don't understand why the names are not accepted after ..

**Reply by oleg68** [reply]
> Then I will install GO. Right now it is not installed at all.

Then you may download three .deb files from https://github.com/GrandOrgue/grandorgue/actions/runs/1244918585 and then install them with apt-get as like as they would be from a release.

> When i make the organ with Mr Forman programm, everything is ok ... i don't understand why the names are not accepted after ..

Linux has a case-sensitive filesystem. It requires that the .organ file would contain all sample file names in the correct case.  
Possible this program is intended for using with windows so it might be not case-sensitive aware.

> Yes but i can't correct 8000 files like that :O

Look for a way of doing it autmatically. I think to update an .organ file is easier than to rename 8000 files.

**Reply by Jean-Gabriel1982** [reply]
Yes, i think this is another subject, but, when i create my organ with V Forman's programm, i first put my sample in a file (like, Go, Pos, Rec, Ped)
Then, i import this file in Mr Forman's file.
From that, i build my organ... check it with the program and if it is ok, export for GrandOrgue.
But when i open the .organ file, i have this warning .. maybe a bug ?

> I think to update an .organ file is easier than to rename 8000 files.

Yes, but the result is the same ... faster but same :D

**Reply by larspalo** [reply]
> But when i open the .organ file, i have this warning .. maybe a bug ?

Maybe. But if so, surely in OrganBuilder not in GO. You must realize that the program of Forman is not guaranteed to in every case possible produce a working ODF for GO. And if there's a warning, I'm certain the error is in the output from Formans program not in the parsing of the .organ file inside of GO.

**Reply by Jean-Gabriel1982** [reply]
> 
> 
> > But when i open the .organ file, i have this warning .. maybe a bug ?
> 
> Maybe. But if so, surely in OrganBuilder not in GO. You must realize that the program of Forman is not guaranteed to in every case possible produce a working ODF for GO. And if there's a warning, I'm certain the error is in the output from Formans program not in the parsing of the .organ file inside of GO.

Yes of course .. my apologies I misspoke .. I was talking about a bug in the output of Mr. Forman's program, no GrandOrgue ..

**Reply by larspalo** [reply]
If you have a way of uploading the .organ file we could inspect it and find out if there's some obvious error that's easy to fix.

**Reply by Jean-Gabriel1982** [reply]
> 
> 
> If you have a way of uploading the .organ file we could inspect it and find out if there's some obvious error that's easy to fix.

Yes, here it is :)
[compose.zip](https://github.com/GrandOrgue/grandorgue/files/7189999/compose.zip)

**Reply by larspalo** [reply]
You'll have to examine the naming of the folders/files in comparison to what is written in the .organ file. Since I don't have access to your file system I cannot see what the folders actually are called but something as simple as a single letter in upper case vs lower case either in the folder name or in file name can trigger this.

I find many of the stop constructs of the included ranks to be very strange indeed. I don't know if this is constructed in such a way that's intended or not, though...

By the way, you'll also have multiple warnings of that the pipe doesn't have any default release, right. All the MaxKeyPressTime=99999 should be changed to MaxKeyPressTime=-1.

I'll examine it more later...

**Reply by larspalo** [reply]
Another issue is that the DispDrawstopCols is set to 18 and GO has a maximum of 12 if you're not using absolute positioning. You'll have to rework the layout in another way to get the .organ file to work.

**Reply by Jean-Gabriel1982** [reply]
> Another issue is that the DispDrawstopCols is set to 18 and GO has a maximum of 12 if you're not using absolute positioning. You'll have to rework the layout in another way to get the .organ file to work.

Yep. I think I have to read carefully GO help to create clean organ …

you’re not afraid if I begin to code ? :D

**Reply by larspalo** [reply]
No, not really. As long as you work on your own fork! ;-) If you want to, and really spend the time necessary, you'll learn!
